container/shim/src/modules/log_ingestor.js

```diff
@@ -41,155 +43,160 @@ const NGINX_LOG_KEYS_MAP = {
+ * then sets a timeout to call itself again after at most 1 minute from the last time it was called. It keeps track of
+ * the last line read in the log file and only reads the lines after that line.
+ * This function can be called manually to force a log ingestor submission (e.g. when node is shutting down).
+ * It also enforces a concurrency limit of 1 to prevent multiple executions of this function from running at the same time.
+ *
+ * @returns {Promise<void>} - resolves once the log ingestor finished current run and is scheduled to run again.
+ */
+export default async function startLogIngestor() {
+  return limitConcurrency(executeLogIngestor);
+}
 
-      const isIPFS = urlObj.pathname.startsWith(IPFS_PREFIX);
-      const isIPNS = urlObj.pathname.startsWith(IPNS_PREFIX);
+async function executeLogIngestor() {
+  // clear timeout timer if it exists (this is to prevent multiple timers from being set)
+  if (executeLogIngestor.timeout) clearTimeout(executeLogIngestor.timeout);
 
-      if (!isIPFS && !isIPNS) {
-        continue;
-      }
+  const startTime = Date.now();
 
-      const {
-        clientAddress,
-        numBytesSent,
-        requestId,
-        localTime,
-        status,
-        requestDuration,
-        range,
-        cacheHit,
-        referrer,
-        userAgent,
-        http3,
-        httpProtocol,
-        url,
-      } = vars;
-
-      const cid = urlObj.pathname.split("/")[2];
-
-      if (cid === TESTING_CID) continue;
-
-      pending.push({
-        cacheHit,
-        clientAddress,
-        localTime,
-        numBytesSent,
-        range,
-        referrer,
-        requestDuration,
-        requestId,
-        userAgent,
-        httpStatusCode: status,
-        // If/when "httpProtocol" eventually contains HTTP/3.0, then
-        // the "http3" key can be removed.
-        httpProtocol: http3 || httpProtocol,
-        url,
-      });
-
-      valid++;
-      if (cacheHit) hits++;
-    }
-    if (valid > 0) {
-      debug(
-        `Parsed ${valid} valid retrievals in ${prettyBytes(read.length)} with hit rate of ${Number(
-          ((hits / valid) * 100).toFixed(2)
-        )}%`
-      );
-    }
-  } else {
-    if (hasRead) {
-      await fh.truncate();
-      await fh.close();
-      hasRead = false;
-      fh = await openFileHandle();
-    }
-  }
-  parseLogsTimer = setTimeout(parseLogs, Math.max(10_000 - valid, 1000));
-}
+  if (await isFileAccessible(LOG_FILE)) {
+    // stream the log file and parse the lines
+    const read = await readlines(LOG_FILE);
 
-async function openFileHandle() {
-  return await fsPromises.open("/usr/src/app/shared/nginx_log/node-access.log", "r+");
-}
+    if (read.lines.length) {
+      const logs = [];
+
+      for (let i = 0; i < read.lines.length; i++) {
+        // skip empty lines
+        if (read.lines[i] === "") continue;
+
+        // parse the line into an object
+        const parsed = parseSingleLine(read.lines[i]);
```

Comment:
These kinds of comments are unnecessary imo. Sometimes code can actually be self documenting. I'm singling out this section but I mean it for the whole PR.

I like comments when they clarify business logic or add additional context. 
I dont like comments when they restate the code, it doesn't add value imo.

So for example instead of 

```js
        // skip empty lines
        if (read.lines[i] === "") continue;
```

you can do
```js
        const isEmptyLine = read.lines[i] === ""
        if (isEmptyLine) continue;
```

and it reads just as well. Except now, if that line of code ever changes, you don't have to update its corresponding comment.

I approve of this PR in its current state, but in the future I'd like to see less comments and more descriptive code. Would love to hear others' thoughts on this

Reply:
Good feedback @guanzo! TBH I just got Github Copilot subscription and went a bit crazy with code comments since it was just too easy to generate them. Your example about self documenting code is spot on, I will keep that in mind in the future 💯